src/plasma/plasma_manager.cc

```diff
@@ -1352,27 +1409,69 @@ ClientConnection *ClientConnection_init(PlasmaManagerState *state,
   /* Create a new data connection context per client. */
   ClientConnection *conn = new ClientConnection();
   conn->manager_state = state;
-  ClientConnection_finish_request(conn);
   conn->fd = client_sock;
   conn->num_return_objects = 0;
-
   conn->ip_addr_port = client_key;
   state->manager_connections[client_key] = conn;
   return conn;
 }
 
+struct TransferSock {
+
+  static TransferSock get_sock(){
+    int sock;
+    sock = bind_inet_sock(0, false);
+
+    struct sockaddr address;
+    socklen_t addrlen = sizeof(address);
+    getsockname(sock, &address, &addrlen);
+    struct sockaddr_in *address_in = (struct sockaddr_in *) &address;
+    int port = htons(address_in->sin_port);
+
+    CHECK(listen(sock, 128) != -1);
+    return TransferSock(sock, port, address_in);
+  };
+
+  int sock;
+  int port;
+  struct sockaddr_in *address_in;
+
+  TransferSock(int sock, int port, sockaddr_in *address_in){
+    this->sock = sock;
+    this->port = port;
+    this->address_in = address_in;
+  }
+
+};
+
 ClientConnection *ClientConnection_listen(event_loop *loop,
                                           int listener_sock,
                                           void *context,
-                                          int events) {
+                                          int events,
+                                          char conn_type) {
+  // SERVER
   PlasmaManagerState *state = (PlasmaManagerState *) context;
   int new_socket = accept_client(listener_sock);
+  LOG_DEBUG("New client connection with fd %d", new_socket);
   char client_key[8];
   snprintf(client_key, sizeof(client_key), "%d", new_socket);
   ClientConnection *conn = ClientConnection_init(state, new_socket, client_key);
 
+  if(conn_type == 'r'){
```

Comment:
As we discussed, I think a lot of the logic here can go away. There are a couple options I think.

1. Have each plasma manager listen on two TCP sockets at the start. One is used for receiving data and one is used for control messages (transfer requests and replies). Each of these two sockets will have an event handler.

2. Use a single socket, but send a "type" along with each message to distinguish the two types.